funtools/man/man7/funview.7
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "funview 7"
.TH funview 7 "April 14, 2011" "version 1.4.5" "SAORD Documentation"
.SH "NAME"
Funview \- Database View Support for Tables
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
This document contains a summary of the options for utilizing
database-inspired Views of tables.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBDatabase Views\fR
.PP
In database parlance, a \fBView\fR defines a \*(L"virtual table\*(R", i.e.,
a description of row and/or column selection filters (but with no
permanent storage space allocated). When used in place of a table, a
View selects the specified rows and/or columns from one or more real
tables. Views enable you to see complicated data tables in a more
convenient format. They also can be used as a security mechanism, by
restricting user access to specific columns and/or rows.  [See:
.PP
http://www.cs.unibo.it/~ciaccia/COURSES/RESOURCES/SQLTutorial/sqlch5.htm
.PP
for a good discussion of \s-1SQL\s0 Views.]
.PP
Funtools supports an expanded notion of Views for all tabular data
(\s-1FITS\s0 tables, raw binary tables, and \s-1ASCII\s0 column files). Funtools
Views allow you to pre-set values for the filter specification, the
columns to activate, and display format (though the latter is for
fundisp only).  Setting the filter and column activation values
provides functionality equivalent to that of a classical database
View, while the ability to set the format is similar to classical
report writing capabilities.
.PP
\&\fBFuntools View Attributes\fR
.PP
A Funtools View is a text file containing one or more of the following
columns:
.PP
.Vb 7
\&  column         description
\&  ------         -----------------------------
\&  view           name of view
\&  file           data file name or template
\&  filter         filter specification
\&  columns        columns to activate
\&  format         fundisp format specification
.Ve
.PP
All of the attribute columns are optional, including
the \fBview\fR name itself. This means that a View can be named or
unnamed. Unnamed Views can refer to a specific file or a template of
files (obviously if neither the view or the file column is specified,
the input View specification will never be used). You can specify any
combination of filter, column, and format parameters. (It also is
possible to apply file-specific View to other files; see the discussion
on \fBView Lists\fR below). Each column has a size limit of 1024 characters.
.PP
For example, consider the following View file:
.PP
.Vb 13
\&  view    file                    format  columns       filter
\&  ----    ----------------------  ------  ------------  -------
\&  x3      ${HOME}/data/snr.ev     I=%4d   x y pi pha    cir 512 512 .1 
\&  x2      ${HOME}/data/snr.ev             x y pi pha    cir 512 512 .1 
\&  x1      ${HOME}/data/snr.ev                           cir 512 512 .1 
\&  x1a     ${HOME}/data/snr.ev             x y pi pha
\&  x0      ${HOME}/data/snr.ev
\&  xf                              I=%4d
\&  xc                                      x y pi pha
\&  xr                                                    cir 512 512 .1
\&          *.ev                            x y pi pha
\&          *.fit                           x y dx dy     cir 400 400  3
\&          *.fits                  I=%3d   x y dx dy     cir 400 400  3
.Ve
.PP
This database example is in rdb format, i.e. using tab delimiters and
permitting null values. Any valid \s-1ASCII\s0 table format is acceptable,
but if you use a format that does not permit null values, it will be
necessary to quote the null strings.
.PP
The first five entries (x3, x2, x1, x1a, x0) are named entries defining
default values specifically for the snr.ev data file. Typically, you
would use these Views by specifying View name, and the corresponding
file, filter, column, and format values would be used. Note that the x0
View is essentially an alias for the pathname of this file.
.PP
The next three entries define defaults that can be applied to any
file.  You typically would use these View names in conjunction with
a specific file name (see \fBView Lists\fR below) so that the associated
parameter(s) were applied to that file.
.PP
The last three entry in the database define unnamed Views that
pertains to all files ending with the specified templates. In these
cases, any View that specifies a file name matching the file template
would be processed with the associated parameter attributes.
.PP
\&\fBInvoking a Funtools View (in Place of an Input File)\fR
.PP
To use a Funtools View, you simply pre-pend the \*(L"v:\*(R" prefix to a View name or
a file name where an input file name usually is specified. For example:
.PP
.Vb 1
\&  fundisp v:x3
.Ve
.PP
specifies that the View named x3 (with its file name and associated
parameters) is processed as the input file to fundisp. Using the
example database, above, this is equivalent to:
.PP
.Vb 1
\&  fundisp  \-f "I=%4d" ${HOME}/data/snr.ev'[cir 512 512 .1]'  "x y pi pha"
.Ve
.PP
That is, the format is used with fundisp's \-f (format) switch, while the
filename and extension are composed of the x3 View's filename and
region filter.
.PP
Similarly, executing a command such as:
.PP
.Vb 1
\&  fundisp v:foo.fit
.Ve
.PP
will match the unnamed View associated with the template \*(L"*.fit\*(R".
This is equivalent to executing:
.PP
.Vb 1
\&  fundisp foo.fit'[cir 400 400 3]' "x y dx dy"
.Ve
.PP
Of course, if you omit the \*(L"v:\*(R" prefix, then no View processing takes place:
.PP
.Vb 2
\&  fundisp foo.fit    # process foo.fit without any View parameters
\&  fundisp x3         # error (assuming there is no file named x3)
.Ve
.PP
\&\fBBasic View Matching Rules\fR
.PP
When a \*(L"v:\*(R" prefix is recognized, Funtools searches for a View database
file in the following order:
.PP
.Vb 5
\&  location             description
\&  ------------         ------------------------------------
\&  FUN_VIEWFILE         environment variable (any file name)
\&  ./.funtools.vu       hidden file, default name
\&  $HOME/.funtools.vu   hidden file, default name
.Ve
.PP
The first View database file located is used to construct a new
filename, as well as an activation column specification and a format
specification. The following rules are used:
.PP
1. An attempt is made to match the input name (i.e., the part of the
input View after the \*(L"v:\*(R" prefix) against the \fBview\fR column value
(if present) of each row in the database. If a match is found, the
values of all non-blank columns are saved for later use.  Also note
that the first match terminates the search: i.e., the order of the
database rows matters.

ChangeLog
2016-01-12 waj

	* upgraded TEA to 3.9
	* numerous compiler fixes for 64bit

2008-11-14  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.10 TAGGED FOR RELEASE ****

	* generic/tkTable.c: Set NO_XSETCLIP universally as Tk 8.5/Xft
	seems to ignore XSetClipRectangles.  This may slow things down a
	little, but draws correctly. [Bug 1805350]

	* UPGRADING.txt (removed): no longer need this old v1 upgrade doc.

	* README.txt: remove refs to mac classic stuff, update notes

	* configure, configure.in, tclconfig/tcl.m4: Update to TEA 3.7
	* tests/tkTable.test, unix/tktable.spec, generic/version.h:
	* demos/loadtable.tcl: Update version to 2.10
	* Makefile.in: update dist target, include library/tktable.py

	* demos/tktable.py: note that this version is outdated
	* library/tktable.py (NEW): Added new version of tktable wrapper
	for Python/Tkinter from GPolo.  Not 100% compatible with previous
	version. [Patch 2244167]

	* generic/tkTableTag.c (Table_TagCmd): remove interp->result set
	work-around for old bug as i->result is no longer exposed in 8.6.

	* generic/tkTable.c:     Use safe Tcl_SetObjResult instead of trying
	* generic/tkTableCell.c: to set existing objResult value.
	* generic/tkTableCmds.c: 

2006-07-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* license.txt (bourbon_ware): Be less restrictive in the
	bourbon_ware clause.

2006-01-25  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure.in, configure, tclconfig/tcl.m4: update to TEA 3.5

2006-01-22  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure: update to TEA 3.4 interim

2005-12-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableCmds.c (Table_ActivateCmd): remove potential
	undefined behavior warning.

2005-12-02  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure.in, configure: update to TEA 3.4

2005-05-11  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (html): switch to groff to fix botched html output.
	* doc/tkTable.html: regen [Bug 1198478]
	* doc/tkTable.n: use '' around \\ items.

2005-03-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (AR): use @AR@, improve html doc target
	* configure, configure.in, tclconfig/tcl.m4: TEA 3.2 patch update

	* doc/tkTable.n, doc/kTable.html: doc cleanup

2005-02-01  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure, configure.in, tclconfig/tcl.m4: update to TEA 3.2

2004-11-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* library/tkTable.tcl (::tk::table::SelectAll): remove invalid ref
	to HandleType proc. [Bug 1068874] (ade)

2004-07-22  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.9 TAGGED FOR RELEASE ****

	* configure, tclconfig/tcl.m4: update tcl.m4 TEA_PATH_X that adds
	Aqua build recognition and TEA_WINDOWINGSYSTEM variable.

2004-07-20  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/version.h, unix/tktable.spec:    updated for v2.9
	* demos/loadtable.tcl, tests/tkTable.test:

	* Makefile.in, configure.in, configure, tclconfig/tcl.m4: update
	to TEA 3.1, use CFLAGS instead of EXTRA_CFLAGS and bump the
	version number of Tktable to 2.9.

	* generic/tkTable.h: add #include <ctype.h>

	* generic/tkTableCell.c (TableGetCellValue, TableSetCellValue): 
	Do faster cache checks when DATA_CACHE is the only data source.
	In fallback from command that errors to array var, retrieve array
	value immediately on current get|set request.
	When setting the value with cache and array sources, don't do the
	caching branch as the array trace will do that already.
	(TableMoveCellValue): Do the efficient cache value management when
	cache is the only data source, otherwise the regular fallback has
	all the necessary checks. [Bug #919872]

	* generic/tkTable.c (TableVarProc): allow setting NULL value in
	cache hash from array trace.

2004-06-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableWin.c (EmbWinDisplay): make min window displayed
	have width/height 1 pixel min (was 3px)

	* generic/tkTable.c (TableDisplay): Don't draw when we don't have
	the space for it. [Bug 747038]

	* doc/tkTable.html, doc/tkTable.n: improve docs to note
	requirement of specifying a data source. [Bug 963204]

2004-02-17  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableCmds.c (Table_ActivateCmd): test implementation
	of an "unactivate" ($table activate "") command to remove the
	active cell. (becroft)

2004-02-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* library/tkTable.tcl (::tk::table::Button1): account for what
	resizeborders are set when selecting cell [Bug 876320] (ferenc) 
	(<Prior> <Next>): activate topleft, not @0,0 (accounts for title
	rows) [Bug 879347] (ferenc)

2003-12-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in: updated to TEA3 spec
	* configure:
	* configure.in:
	* generic/tkTable.c:
	* generic/tkTableInitScript.h:
	* generic/version.h:
	* mac/mac_tkTable.r:
	* tclconfig/tcl.m4:

	* win/makefile.vc: minor updates, probably still out of date

	* tclconfig/ChangeLog (removed): not necessary

	* generic/tkTable.c: added Tk_ClassProcs and TableWorldChanged
	proc if built against 8.4

	* generic/tkTable.h: add 8.4 ifdef, remove ANSI_ARGS usage


funtools/funview.c
/*
 *	Copyright (c) 2005 Smithsonian Astrophysical Observatory
 */

#include <funtoolsP.h>

/*
 *
 * private routines
 *
 */

/* default data base files */
#define FUN_VIEWFILE ".funtools.vu"
#define FUN_VIEWHOME "$HOME/.funtools.vu"

/* default text options for database open */
#define FUN_VIEWEXT "[1,TEXT(alen=1024)]"

/* default is to allow multiple matches */
#define FUN_VIEW_MATCH_DEFAULT 1

/* if a list is specified, do we also match templates? */
#define FUN_VIEW_DOROW_DEFAULT 1

/* types of matches */
#define MATCH_VIEW 1
#define MATCH_FILE 2
#define MATCH_VTMP 4
#define MATCH_FTMP 8

/* this should match what is in fun_viewopen's alen above */
#define SZ_VIEW 1024

/* parser characteristics struct */
typedef struct _rowstruct{
  int type;
  char view[SZ_VIEW];
  char file[SZ_VIEW];
  char fmt[SZ_VIEW];
  char cols[SZ_VIEW];
  char filt[SZ_VIEW];
} *Row, RowRec;

typedef struct _viewstruct{
  char *vname;
  char *vroot;
  int nlist;
  int maxlist;
  char **vlist;
  Fun ifun;
  Fun vfun;
  int nrow;
  int maxrow;
  Row *rows;
  Row file, filt, cols, fmt;
  Row trow;
} *View, ViewRec;

#ifdef ANSI_FUNC
static int
pathcmp(char *s1, char *s2)
#else
static int
pathcmp(s1, s2)
  char *s1;
  char *s2;
#endif
{
  char tbuf1[SZ_LINE];
  char tbuf2[SZ_LINE];
  char tbuf1a[SZ_LINE];
  char tbuf2a[SZ_LINE];

  /* resolve pathnames and expand environment variables */
  ExpandEnv(s1, tbuf1, SZ_LINE);
  ExpandEnv(s2, tbuf2, SZ_LINE);
  ResolvePath(tbuf1, tbuf1a, SZ_LINE);
  ResolvePath(tbuf2, tbuf2a, SZ_LINE);

  /* now compare strings */
  return strcmp(tbuf1a, tbuf2a);
}

#ifdef ANSI_FUNC
static int
ViewFree(View view)
#else
static int ViewFree(view)
     View view;
#endif
{
  int i;

  /* sanity check */
  if( !view ) return 0;

  /* free up resources */
  if( view->vfun )  FunClose(view->vfun);
  if( view->vname ) xfree(view->vname);
  if( view->vroot ) xfree(view->vroot);
  if( view->trow )  xfree(view->trow);
  if( view->rows ){
    for(i=0; i<view->nrow; i++){
      if( view->rows[i] ) xfree(view->rows[i]);
    }
    xfree(view->rows);
  }
  if( view->vlist ){
    for(i=0; i<view->nlist; i++){
      if( view->vlist[i] ) xfree(view->vlist[i]);
    }
    xfree(view->vlist);
  }
  xfree(view);
  return 1;
}

#ifdef ANSI_FUNC
static View
ViewNew(Fun fun, char *vname, char *vptr, char *vlist, char *dbname, int dbmax)
#else
static View
ViewNew(fun, vname, vptr, vlist, dbname, dbmax)
     Fun fun;
     char *vname;
     char *vptr;
     char *vlist;
     char *dbname;
     int dbmax;
#endif
{
  int ip=0;
  char *s, *t;
  char *rmode="r";
  char *fmode="rV";
  char *params=NULL;
  char tbuf[SZ_LINE];
  char dtype[SZ_LINE];
  View view=NULL;

  /* look for view file in the usual places */
  if( (s=(char *)getenv("FUN_VIEWFILE")) && (t=Access(s, rmode)) ){
    xfree(t);
    goto gotv;
  }
  else if( (t=Access(FUN_VIEWFILE, rmode)) ){
    xfree(t);
    s = FUN_VIEWFILE;
    goto gotv;
  }
  else if( (t=Access(FUN_VIEWHOME, rmode)) ){
    xfree(t);
    s = FUN_VIEWHOME;
    goto gotv;
  }
  else{
    return NULL;
  }

gotv:
  /* allocate the view struct */
  if( !(view=(View)xcalloc(1, sizeof(ViewRec))) ){
    gerror(stderr, "can't allocate view database struct\n");
    return NULL;
  }
  /* allocate temp row */
  if( !(view->trow=(Row)xcalloc(1, sizeof(RowRec))) ){
    gerror(stderr, "can't allocate view temp row struct\n");
    ViewFree(view);
    return NULL;
  }

  /* add default text options to the filename, if needed */
  if( strchr(s, '[') ){
    strncpy(tbuf, s, SZ_LINE);
  }
  else{
    snprintf(tbuf, SZ_LINE, "%s%s", s, FUN_VIEWEXT);
  }

  /* open the file */
  if( !(view->vfun=FunOpen(tbuf, fmode, NULL)) ){
    gerror(stderr, "can't open view database file: %s\n", s);
    ViewFree(view);
    return NULL;
  }

  /* select columns from view database */
  snprintf(dtype, SZ_LINE-1, "%dA", SZ_VIEW);
  FunColumnSelect(view->vfun, sizeof(RowRec), params,
		  "view",    dtype, rmode, FUN_OFFSET(Row, view),
		  "file",    dtype, rmode, FUN_OFFSET(Row, file),
		  "format",  dtype, rmode, FUN_OFFSET(Row, fmt),
		  "columns", dtype, rmode, FUN_OFFSET(Row, cols),
		  "filter",  dtype, rmode, FUN_OFFSET(Row, filt),
		  NULL);

  /* save original view string */
  view->vname = xstrdup(vname);

  /* save view name without any bracket extension */
  view->vroot = xstrdup(vptr);
  if( (s=strchr(view->vroot, '[')) ) *s = '\0';

  /* save original input Fun handle */
  view->ifun = fun;

  /* save vlist as separate strings */
  if( vlist ){
    newdtable(",");
    while( word(vlist, tbuf, &ip) ){
      if( view->nlist >= view->maxlist ){
	view->maxlist += SZ_VIEW;
	if( view->vlist )
	  view->vlist = (char **)xrealloc(view->vlist,
					  view->maxlist*sizeof(char **));
	else
	  view->vlist = (char **)xmalloc(view->maxlist*sizeof(char **));
      }
      /* add this record to the array of matches */
      view->vlist[view->nlist++] = xstrdup(tbuf);
    }
    freedtable();
  }

  /* temp row, in case template is matched as a file => use input file name */
  strncpy(view->trow->file, vptr, SZ_VIEW-1);
  if( (s=strrchr(view->trow->file, '[')) ){
    /* remove extension from file (we'll put it back later, if necessary) */
    *s = '\0';
  }
  /* return name of database used, if necessary */
  if( dbname ){
    FunInfoGet(view->vfun, FUN_FNAME, &s, 0);
    strncpy(dbname, s, dbmax-1);
  }

  /* return the good news */
  return view;
}

#ifdef ANSI_FUNC
static int
ViewMatchRow(View view, Row row)
#else
static int
ViewMatchRow(view, row)
     View view;
     Row row;
#endif
{
  char *root=NULL;

  /* sanity check */
  if( !row || !view->vroot || !*view->vroot ) return 0;

  /* root is file without path */
  if( (root=strrchr(row->file, '/')) )
    root++;
  else
    root=row->file;

  /* look for various types of match */
  if( *row->view && !strcmp(view->vroot, row->view) )
    row->type |= MATCH_VIEW;
  /* if not a vew, perhaps a view template */
  else if( *row->view && tmatch(view->vroot, row->view) )
    row->type |= MATCH_VTMP;
  /* check for file, with and without path */
  if( (*row->file &&!pathcmp(view->vroot, row->file)) || 
      (root && *root && !strcmp(view->vroot, root)))
    row->type |= MATCH_FILE;
  /* if not a file, perhaps a file template */
  else if( root && *root && tmatch(view->vroot, root) )
    row->type |= MATCH_FTMP;

  /* return the results */
  return row->type;
}

#ifdef ANSI_FUNC
static int
ViewSaveRow(View view, Row row)
#else
static int
ViewSaveRow(view, row)
     View view;
     Row row;
#endif
{
  /* make sure we have enough space */
  if( view->nrow >= view->maxrow ){
    view->maxrow += SZ_VIEW;
    if( view->rows )
      view->rows = (Row *)xrealloc(view->rows, view->maxrow*sizeof(Row));
    else
      view->rows = (Row *)xmalloc(view->maxrow*sizeof(Row));
  }
  /* add this record to the array of matches */
  view->rows[view->nrow++] = row;
  return view->nrow;
}

#ifdef ANSI_FUNC
static int
ViewMatchList(View view, Row row, int dorow)
#else
static int
ViewMatchList(view, row, dorow)
     View view;
     Row row;
     int dorow;
#endif
{
  int i;

  /* sanity check */
  if( !row || !view->vlist ) return 0;

  /* look though all view strings in the list */
  for(i=0; i<view->nlist; i++){
    /* does list view match row view? */
    if( !strcmp(view->vlist[i], row->view) ){
      row->type |= MATCH_VIEW;
      break;
    }
  }

  /* also check row matches, if necessary */
  if( dorow ) ViewMatchRow(view, row);

  /* return the results */
  return row->type;
}

#ifdef ANSI_FUNC
static int
ViewProcessMatches(View view, char *fname, int fmax)
#else
static int
ViewProcessMatches(view, fname, fmax)
     View view;
     char *fname;
     int fmax;
#endif
{
  int i;
  int nv=0;
  int nf=0;
  int nvt=0;
  int nft=0;
  int fgot=-1;
  int tgot=-1;
  int domulti=FUN_VIEW_MATCH_DEFAULT;
  char *s=NULL;

  /* sanity check */
  if( !view ) return 0;

  /* do we match a single row or use multiple row to complete the spec? */
  if( (s=getenv("FUN_VIEWMATCH")) ){
    if( !strncmp(s, "m", 1) ){
      domulti = 1;
    }
    else if( !strncmp(s, "s", 1) ){
      domulti = 0;
    }
  }

  /* collect separate parts of fname using increasingly general matching */
  if( view->nrow ){
    /* get match counts */
    for(i=0; i<view->nrow; i++){
      if( view->rows[i]->type & MATCH_VIEW )
	nv++;
      if( view->rows[i]->type & MATCH_VTMP )
	nvt++;
      if( view->rows[i]->type & MATCH_FILE )
	nf++;
      if( view->rows[i]->type & MATCH_FTMP )
	nft++;
    }
    /* view matches are the most desirable */
    if( nv ){
      for(i=0; i<view->nrow; i++){
	if( view->rows[i]->type & MATCH_VIEW ){
	  /* if we are using a list, we always take the original filename */
	  if( view->nlist )
	    view->file = view->trow;
	  else if( !view->file && *view->rows[i]->file )
	    view->file = view->rows[i];
	  if( !view->filt && *view->rows[i]->filt )
	    view->filt = view->rows[i];
	  if( !view->cols && *view->rows[i]->cols )
	    view->cols = view->rows[i];
	  if( !view->fmt  && *view->rows[i]->fmt  )
	    view->fmt  = view->rows[i];
	  tgot = i;
	  /* we only use one, unless we are running off a list */
	  if( !view->nlist ) break;
	}
      }
    }
    /* if no view matches, try file matches */
    else if( nf ){
      for(i=0; i<view->nrow; i++){
	if( view->rows[i]->type & MATCH_FILE ){
	  if( !view->file && *view->rows[i]->file )
	    view->file = view->rows[i];
	  if( !view->filt && *view->rows[i]->filt )
	    view->filt = view->rows[i];
	  if( !view->cols && *view->rows[i]->cols )
	    view->cols = view->rows[i];
	  if( !view->fmt  && *view->rows[i]->fmt  )
	    view->fmt  = view->rows[i];
	  fgot = i;
	  break;
	}
      }
    }
    /* if we got a match, only do templates if multiple matches is true */
    if( (tgot==-1 && fgot==-1) || domulti ){
      for(i=0; i<view->nrow; i++){
	if( tgot == i ) continue;
	if( view->rows[i]->type & MATCH_VTMP ){
	  /* use the input file name in this case */
	  if( !view->file && *view->rows[i]->file )
	    view->file = view->trow;
	  if( !view->filt && *view->rows[i]->filt )
	    view->filt = view->rows[i];
	  if( !view->cols && *view->rows[i]->cols )
	    view->cols = view->rows[i];
	  if( !view->fmt  && *view->rows[i]->fmt  )
	    view->fmt  = view->rows[i];
	  /* mark use of view template so we don't do file templates */
	  tgot = i;
	  /* stop at one if not domulti */
	  if( !domulti ) break;
	}
      }
    }
    /* if we got a match, only do templates if multiple matches is true */
    if( (tgot==-1 && fgot==-1) || domulti ){
      for(i=0; i<view->nrow; i++){
	if( fgot == i ) continue;
	if( view->rows[i]->type & MATCH_FTMP ){
	  /* use the input file name in this case */
	  if( !view->file && *view->rows[i]->file )
	    view->file = view->trow;
	  if( !view->filt && *view->rows[i]->filt )
	    view->filt = view->rows[i];
	  if( !view->cols && *view->rows[i]->cols )
	    view->cols = view->rows[i];
	  if( !view->fmt  && *view->rows[i]->fmt  )
	    view->fmt  = view->rows[i];
	  /* stop at one if not domulti */
	  if( !domulti ) break;
	}
      }
    }

    /* now we can make up the fname from the pieces we have found */
    if( view->file && *view->file->file ){
      /* see if original input had an extension */
      if( (s=strchr(view->vname, '[')) ){
	/* if we can guess that original was a filter, don't add new one */
	if(  strstr(s,"ann") ||
	     strstr(s,"box") ||
	     strstr(s,"cir") ||
	     strstr(s,"ell") ||
	     strstr(s,"lin") ||
	     strstr(s,"pan") ||
	     strstr(s,"pie") ||
	     strstr(s,"poi") ||
	     strstr(s,"qtp") ||
	     strstr(s,"pol") ||
	     strstr(s,"fie") ||
	     strstr(s,"bpa") ||
	     strstr(s,"cpa") ||
	     strstr(s,"epa") ||
	     /* guess at some sort of filter operation */
	     strpbrk(s, "!&|~=<>+-/%^") ){
	  snprintf(fname, fmax, "%s%s", view->file->file, s);
	}
	/* otherwise add original + new filter, if we have one */
	else{
	  if( view->filt && *view->filt->filt ){
	    snprintf(fname, fmax, "%s%s[%s]",
		     view->file->file, s, view->filt->filt);
	  }
	  /* just add original */
	  else{
	    snprintf(fname, fmax, "%s%s", view->file->file, s);
	  }
	}
      }
      /* no original filter */
      else{
	/* otherwise new filter, if we have one */
	if( view->filt && *view->filt->filt ){
	  snprintf(fname, fmax, "%s[%s]", view->file->file, view->filt->filt);
	}
	/* just the file name */
	else{
	  snprintf(fname, fmax, "%s", view->file->file);
	}
      }
      /* save other view info */
      if( view->ifun ){
	if( view->cols && *view->cols->cols ) 
	  view->ifun->vcols = xstrdup(view->cols->cols);
	if( view->fmt  && *view->fmt->fmt )
	  view->ifun->vfmt  = xstrdup(view->fmt->fmt);
      }
    }
    return 1;
  }
  else{
    return 0;
  }
}

/*
 *
 * public routines
 *
 */

/*
 *
 * FunView -- translate view into a filename + view params
 *
 */
#ifdef ANSI_FUNC
int
FunView(Fun fun, char *vname, char *vmode, char *fname, int fmax)
#else
int
FunView(fun, vname, vmode, fname, fmax)
     Fun fun;
     char *vname;
     char *vmode;
     char *fname;
     int fmax;
#endif
{
  int nev=0;
  int got=0;
  int dorow=FUN_VIEW_DOROW_DEFAULT;
  char *s=NULL;
  char *vptr=NULL;
  char vlist[SZ_LINE];
  Row row=NULL;
  View view=NULL;

  /* sanity checks */
  if( !fname || !fmax ) return 0;
  *fname = '\0';
  if( !vname || !*vname ) return 0;

  /* assume we don't have a view */
  strncpy(fname, vname, fmax-1);
  fname[fmax-1] = '\0';

  /* sanity check -- this prevents infinite recusion below */
  if( !fun || (vmode && strchr(vmode, 'V')) ) return 0;

  /* see if a view ("v:...") was specified */
  memset(vlist, 0, SZ_LINE);
  vptr = vname;
  if( !*vptr || (*vptr != 'v') ) return 0;
  vptr++;
  if( !*vptr || (*vptr != ':') ) return 0;
  vptr++;
  if( !*vptr ) return 0;
  /* v::vid1,vid2...:file... */
  if( (*vptr == ':') || (*vptr == '+') || (*vptr == '-') ){
    if( *vptr == '+' ) dorow = 1;
    else if( *vptr == '-' ) dorow = 0;
    vptr++;
    /* must have a closing ':' */
    if( !*vptr  || !(s = strchr(vptr, ':')) ) return 0;
    /* copy the list */
    strncpy(vlist, vptr, MIN(s-vptr,SZ_LINE-1));
    /* bump past the ':' */
    vptr = s+1;
  }
  if( !*vptr ) return 0;

  /* now assume we don't match and want to use the supplied filename */
  strncpy(fname, vptr, fmax-1);
  fname[fmax-1] = '\0';

  /* open view file */
  if( !(view=ViewNew(fun, vname, vptr, vlist, NULL, 0)) ) goto done;

  /* read rows from view database */
  while( (row=FunTableRowGet(view->vfun, NULL, 1, NULL, &nev)) && nev ){
    /* look for a match */
    if( view->vlist ){
      if( ViewMatchList(view, row, dorow) ){
	ViewSaveRow(view, row);
	/* clear row */
	row = NULL;
      }
    }
    else{
      if( ViewMatchRow(view, row) ){
	ViewSaveRow(view, row);
	/* clear row */
	row = NULL;
      }
    }
    /* free record if it was not saved (and cleared) */
    if( row ){
      xfree(row);
      row = NULL;
    }
  }

  /* process view information and generate a new file name */
  got = ViewProcessMatches(view, fname, fmax);

done:
  /* clean up */
  ViewFree(view);
  if( row ) xfree(row);

  /* return the news */
  return got;
}
